In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt
#import seaborn as sns
#from sklearn import preprocessing
#color = sns.color_palette()

%matplotlib inline

pd.options.mode.chained_assignment = None  # default='warn'
pd.options.display.max_columns = 999

In [31]:
train_df = pd.read_csv("../data/train.csv")
test_df = pd.read_csv("../data/test.csv")
print("Train shape : ", train_df.shape)
print("Test shape : ", test_df.shape)

('Train shape : ', (4209, 378))
('Test shape : ', (4209, 377))


In [3]:
train_df.head()

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  X10  X11  X12  X13  X14  X15  X16  \
0   0  130.81   k  v  at  a  d  u  j  o    0    0    0    1    0    0    0   
1   6   88.53   k  t  av  e  d  y  l  o    0    0    0    0    0    0    0   
2   7   76.26  az  w   n  c  d  x  j  x    0    0    0    0    0    0    0   
3   9   80.62  az  t   n  f  d  x  l  e    0    0    0    0    0    0    0   
4  13   78.02  az  v   n  f  d  h  d  n    0    0    0    0    0    0    0   

   X17  X18  X19  X20  X21  X22  X23  X24  X26  X27  X28  X29  X30  X31  X32  \
0    0    1    0    0    1    0    0    0    0    0    0    0    0    1    0   
1    0    1    0    0    0    0    0    0    0    1    0    0    0    1    0   
2    1    0    0    0    0    0    0    0    0    1    1    1    0    1    0   
3    0    0    0    0    0    0    0    0    0    1    1    1    0    1    0   
4    0    0    0    0    0    0    0    0    0    1    1    1    0    1    0   

   X33  X34  X35  X36  X37  X38  X39  X40  X41  X42  X43  X44  X45  X46  X47  \
0    0    0    1    0    1    0    0    0    0    0    0    0    0    1    0   
1    0    0    1    0    1    0    0    0    0    0    0    0    0    0    0   
2    0    0    1    0    1    0    0    0    0    0    1    0    0    1    0   
3    0    0    1    0    1    0    0    0    0    0    1    0    0    1    0   
4    0    0    1    0    1    0    0    0    0    0    1    0    0    1    0   

   X48  X49  X50  X51  X52  X53  X54  X55  X56  X57  X58  X59  X60  X61  X62  \
0    0    0    0    0    0    0    0    0    0    0    1    0    0    0    0   
1    0    0    0    1    0    0    0    0    0    0    0    0    0    1    0   
2    0    0    0    1    0    0    1    0    0    0    1    0    0    1    0   
3    0    0    0    0    0    0    1    0    0    0    0    0    0    1    0   
4    0    0    0    1    0    0    1    0    0    0    0    0    0    1    0   

   X63  X64  X65  X66  X67  X68  X69  X70  X71  X73  X74  X75  X76  X77  X78  \
0    0    0    0    0    0    1    0    1    0    0    1    0    0    0    0   
1    0    0    0    0    0    0    0    1    0    0    1    0    0    0    0   
2    0    0    0    0    0    1    0    1    1    0    1    1    1    0    0   
3    0    0    0    0    0    0    0    1    1    0    1    0    1    0    0   
4    0    0    0    0    0    0    0    1    0    0    1    0    1    0    0   

   X79  X80  X81  X82  X83  X84  X85  X86  X87  X88  X89  X90  X91  X92  X93  \
0    0    0    0    0    0    0    1    0    0    0    0    0    0    0    0   
1    0    1    0    0    0    0    1    0    0    0    0    0    0    0    0   
2    0    1    0    0    0    1    1    0    0    0    0    0    0    0    0   
3    0    1    0    0    0    1    0    0    0    0    0    0    0    0    0   
4    0    1    0    0    0    0    0    0    0    0    0    0    0    0    0   

   X94  X95  X96  X97  X98  X99  X100  X101  X102  X103  X104  X105  X106  \
0    0    0    0    0    0    0     0     0     0     0     0     0     0   
1    0    0    1    0    1    0     1     1     0     0     0     0     0   
2    0    0    1    0    1    0     0     1     0     0     0     0     0   
3    0    0    1    0    1    0     0     1     0     0     0     0     0   
4    0    0    1    0    1    0     0     1     0     0     0     0     0   

   X107  X108  X109  X110  X111  X112  X113  X114  X115  X116  X117  X118  \
0     0     0     0     0     1     0     0     1     0     1     0     1   
1     0     0     0     0     1     0     0     0     0     0     0     1   
2     0     0     0     0     1     0     0     0     0     0     0     0   
3     0     1     0     0     1     0     0     1     0     0     0     0   
4     0     1     0     0     1     0     0     1     0     0     0     0   

   X119  X120  X122  X123  X124  X125  X126  X127  X128  X129  X130  X131  \
0     1     1     0     0     0     0     0     0     1     0     0     1   
1     1     1     0     0     0     0     0     1

In [4]:
dtype_df = train_df.dtypes.reset_index()
dtype_df.columns = ["Count", "Column Type"]
dtype_df.groupby("Column Type").aggregate('count').reset_index()

Column Type  Count
0       int64    369
1     float64      1
2      object      8

## Feature analysis

There are:
 * 369 ints = actually binary values
 * 1 float = target variable
 * 8 categorical features

## Feature engineering

1. Change categorical variables into binary with one-hot encoding
2. Reduce number of features

### 1. Categorical to one-hot:

In [32]:
def to_one_hot(column, df):
    """
    Replaces a given column with N new one-hot encoded columns.
    N = number of unique values of given column.
    """
    
    # Get unique values of column
    unique_vals = df[column].unique()
    
    # Add a new zero-filled column for each unique value
    for val in unique_vals:
        new_column_name = column + val
        df[new_column_name] = 0
        
    # Put 1s to new columns:
    # e.g. original df: [1, 'X0'] = 'k' => new df: [1, 'X0k'] = 1
    row_idx = 0
    for categ_val in df[column].values:
        column_name = column + categ_val
        df.set_value(row_idx, column_name, 1)
        row_idx += 1
        
    # Remove the column with categorical values
    df.drop(column, axis=1, inplace=True)
    
    print("Replaced categorical column {} with {} new columns.".format(column, len(unique_vals)))
    
    return df

In [33]:
def all_to_one_hot(df):
    """
    Replaces all categorical columns by one-hot encoded columns.
    See: to_one_hot(column, df)
    """
    
    # Get a list of categorical column names
    categorical_columns = list(train_df.select_dtypes(include=['object']))
    
    print("These {} categorical columns will be replaced: {}".format(len(categorical_columns), categorical_columns))
    
    for column in categorical_columns:
        df = to_one_hot(column, df)
    
    print("New shape of dataframe after all replacements: ", df.shape)
    return df

In [34]:
train_df = all_to_one_hot(train_df)
train_df.head()

These 8 categorical columns will be replaced: ['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8']
Replaced categorical column X0 with 47 new columns.
Replaced categorical column X1 with 27 new columns.
Replaced categorical column X2 with 44 new columns.
Replaced categorical column X3 with 7 new columns.
Replaced categorical column X4 with 4 new columns.
Replaced categorical column X5 with 29 new columns.
Replaced categorical column X6 with 12 new columns.
Replaced categorical column X8 with 25 new columns.
('New shape of dataframe after all replacements: ', (4209, 565))


ID       y  X10  X11  X12  X13  X14  X15  X16  X17  X18  X19  X20  X21  \
0   0  130.81    0    0    0    1    0    0    0    0    1    0    0    1   
1   6   88.53    0    0    0    0    0    0    0    0    1    0    0    0   
2   7   76.26    0    0    0    0    0    0    0    1    0    0    0    0   
3   9   80.62    0    0    0    0    0    0    0    0    0    0    0    0   
4  13   78.02    0    0    0    0    0    0    0    0    0    0    0    0   

   X22  X23  X24  X26  X27  X28  X29  X30  X31  X32  X33  X34  X35  X36  X37  \
0    0    0    0    0    0    0    0    0    1    0    0    0    1    0    1   
1    0    0    0    0    1    0    0    0    1    0    0    0    1    0    1   
2    0    0    0    0    1    1    1    0    1    0    0    0    1    0    1   
3    0    0    0    0    1    1    1    0    1    0    0    0    1    0    1   
4    0    0    0    0    1    1    1    0    1    0    0    0    1    0    1   

   X38  X39  X40  X41  X42  X43  X44  X45  X46  X47  X48  X49  X50  X51  X52  \
0    0    0    0    0    0    0    0    0    1    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    1    0   
2    0    0    0    0    0    1    0    0    1    0    0    0    0    1    0   
3    0    0    0    0    0    1    0    0    1    0    0    0    0    0    0   
4    0    0    0    0    0    1    0    0    1    0    0    0    0    1    0   

   X53  X54  X55  X56  X57  X58  X59  X60  X61  X62  X63  X64  X65  X66  X67  \
0    0    0    0    0    0    1    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    1    0    0    0    0    0    0   
2    0    1    0    0    0    1    0    0    1    0    0    0    0    0    0   
3    0    1    0    0    0    0    0    0    1    0    0    0    0    0    0   
4    0    1    0    0    0    0    0    0    1    0    0    0    0    0    0   

   X68  X69  X70  X71  X73  X74  X75  X76  X77  X78  X79  X80  X81  X82  X83  \
0    1    0    1    0    0    1    0    0    0    0    0    0    0    0    0   
1    0    0    1    0    0    1    0    0    0    0    0    1    0    0    0   
2    1    0    1    1    0    1    1    1    0    0    0    1    0    0    0   
3    0    0    1    1    0    1    0    1    0    0    0    1    0    0    0   
4    0    0    1    0    0    1    0    1    0    0    0    1    0    0    0   

   X84  X85  X86  X87  X88  X89  X90  X91  X92  X93  X94  X95  X96  X97  X98  \
0    0    1    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    1    0    0    0    0    0    0    0    0    0    0    1    0    1   
2    1    1    0    0    0    0    0    0    0    0    0    0    1    0    1   
3    1    0    0    0    0    0    0    0    0    0    0    0    1    0    1   
4    0    0    0    0    0    0    0    0    0    0    0    0    1    0    1   

   X99  X100  X101  X102  X103  X104  X105  X106  X107  X108  X109  X110  \
0    0     0     0     0     0     0     0     0     0     0     0     0   
1    0     1     1     0     0     0     0     0     0     0     0     0   
2    0     0     1     0     0     0     0     0     0     0     0     0   
3    0     0     1     0     0     0     0     0     0     1     0     0   
4    0     0     1     0     0     0     0     0     0     1     0     0   

   X111  X112  X113  X114  X115  X116  X117  X118  X119  X120  X122  X123  \
0     1     0     0     1     0     1     0     1     1     1     0     0   
1     1     0     0     0     0     0     0     1     1     1     0     0   
2     1     0     0     0     0     0     0     0     0     1     0     0   
3     1     0     0     1     0     0     0     0     0     1     0     0   
4     1     0     0     1     0     0     0     0     0     1     0     0   

   X124  X125  X126  X127  X128  X129  X130  X131  X132  X133  X134  X135  \
0     0     0     0     0     1     0     0     1     0     0     0     0   
1     0     0     0     1     1     0     0     0     1     0

### 2. Dimensionality reduction:

 * analyze column values and discard the ones with only 1 value

In [35]:
def get_constant_columns(df):
    constant_columns = []
    for column in list(df):
        if len(df[column].unique()) == 1:
            constant_columns.append(column)
    print("Constant columns ({}): {}".format(len(constant_columns), constant_columns))
    return constant_columns

In [36]:
constant_columns = get_constant_columns(train_df)

Constant columns (12): ['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290', 'X293', 'X297', 'X330', 'X347']


In [37]:
def drop_columns(df, columns):
    for column in columns:
        df.drop(column, axis=1, inplace=True)
    print("Dropped {} columns: {}".format(len(columns), columns))
    return df    

In [38]:
train_df = drop_columns(train_df, constant_columns)
train_df.head()

Dropped 12 columns: ['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290', 'X293', 'X297', 'X330', 'X347']


ID       y  X10  X12  X13  X14  X15  X16  X17  X18  X19  X20  X21  X22  \
0   0  130.81    0    0    1    0    0    0    0    1    0    0    1    0   
1   6   88.53    0    0    0    0    0    0    0    1    0    0    0    0   
2   7   76.26    0    0    0    0    0    0    1    0    0    0    0    0   
3   9   80.62    0    0    0    0    0    0    0    0    0    0    0    0   
4  13   78.02    0    0    0    0    0    0    0    0    0    0    0    0   

   X23  X24  X26  X27  X28  X29  X30  X31  X32  X33  X34  X35  X36  X37  X38  \
0    0    0    0    0    0    0    0    1    0    0    0    1    0    1    0   
1    0    0    0    1    0    0    0    1    0    0    0    1    0    1    0   
2    0    0    0    1    1    1    0    1    0    0    0    1    0    1    0   
3    0    0    0    1    1    1    0    1    0    0    0    1    0    1    0   
4    0    0    0    1    1    1    0    1    0    0    0    1    0    1    0   

   X39  X40  X41  X42  X43  X44  X45  X46  X47  X48  X49  X50  X51  X52  X53  \
0    0    0    0    0    0    0    0    1    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    1    0    0   
2    0    0    0    0    1    0    0    1    0    0    0    0    1    0    0   
3    0    0    0    0    1    0    0    1    0    0    0    0    0    0    0   
4    0    0    0    0    1    0    0    1    0    0    0    0    1    0    0   

   X54  X55  X56  X57  X58  X59  X60  X61  X62  X63  X64  X65  X66  X67  X68  \
0    0    0    0    0    1    0    0    0    0    0    0    0    0    0    1   
1    0    0    0    0    0    0    0    1    0    0    0    0    0    0    0   
2    1    0    0    0    1    0    0    1    0    0    0    0    0    0    1   
3    1    0    0    0    0    0    0    1    0    0    0    0    0    0    0   
4    1    0    0    0    0    0    0    1    0    0    0    0    0    0    0   

   X69  X70  X71  X73  X74  X75  X76  X77  X78  X79  X80  X81  X82  X83  X84  \
0    0    1    0    0    1    0    0    0    0    0    0    0    0    0    0   
1    0    1    0    0    1    0    0    0    0    0    1    0    0    0    0   
2    0    1    1    0    1    1    1    0    0    0    1    0    0    0    1   
3    0    1    1    0    1    0    1    0    0    0    1    0    0    0    1   
4    0    1    0    0    1    0    1    0    0    0    1    0    0    0    0   

   X85  X86  X87  X88  X89  X90  X91  X92  X94  X95  X96  X97  X98  X99  X100  \
0    1    0    0    0    0    0    0    0    0    0    0    0    0    0     0   
1    1    0    0    0    0    0    0    0    0    0    1    0    1    0     1   
2    1    0    0    0    0    0    0    0    0    0    1    0    1    0     0   
3    0    0    0    0    0    0    0    0    0    0    1    0    1    0     0   
4    0    0    0    0    0    0    0    0    0    0    1    0    1    0     0   

   X101  X102  X103  X104  X105  X106  X108  X109  X110  X111  X112  X113  \
0     0     0     0     0     0     0     0     0     0     1     0     0   
1     1     0     0     0     0     0     0     0     0     1     0     0   
2     1     0     0     0     0     0     0     0     0     1     0     0   
3     1     0     0     0     0     0     1     0     0     1     0     0   
4     1     0     0     0     0     0     1     0     0     1     0     0   

   X114  X115  X116  X117  X118  X119  X120  X122  X123  X124  X125  X126  \
0     1     0     1     0     1     1     1     0     0     0     0     0   
1     0     0     0     0     1     1     1     0     0     0     0     0   
2     0     0     0     0     0     0     1     0     0     0     0     0   
3     1     0     0     0     0     0     1     0     0     0     0     0   
4     1     0     0     0     0     0     1     0     0     0     0     0   

   X127  X128  X129  X130  X131  X132  X133  X134  X135  X136  X137  X138  \
0     0     1     0     0     1     0     0     0     0     1     1     0   
1     1     1     0     0     0     1     0     0